synapse/storage/databases/main/event_federation.py

```diff
@@ -1292,6 +1292,99 @@ def _get_backfill_events(
 
         return event_id_results
 
+    @trace
+    async def record_event_failed_backfill_attempt(
+        self, room_id: str, event_id: str
+    ) -> None:
+        """
+        Record when we fail to backfill an event.
+
+        This information allows us to be more intelligent when we decide to
+        retry (we don't need to fail over and over) and we can process that
+        event in the background so we don't block on it each time.
+
+        Args:
+            room_id: The room where the event failed to backfill from
+            event_id: The event that failed to be backfilled
+        """
+        if (
+            self.database_engine.can_native_upsert
+            and "event_failed_backfill_attempts"
+            not in self.db_pool._unsafe_to_upsert_tables
```

Comment:
`_unsafe_to_upsert_tables` is statically populated. This can never be true.

Reply:
`_unsafe_to_upsert_tables` could change in the future though 🤷, see https://github.com/matrix-org/synapse/pull/13589#discussion_r960436114

Reply:
hrm, I'm not sure I agree with that. The only reason things get added to `_unsafe_to_upsert_tables` is if they were created incorrectly in the first place, and need to have a unique index added later. That's unlikely to happen here.

Reply:
fair enough!

Reply:
Removed with the rest of the emulated upsert code, see https://github.com/matrix-org/synapse/pull/13589#discussion_r961101746